Merge Conflict Markers - The Explicit Pathing, Layering, Iconing, Warning.

san7890 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Hey there,

This PR corrects an issue I've been having with mapmerge2 these last few months. Basically, what it does is create a base `/obj` that is given the name `---Merge Conflict Marker---`. This is fine and all, but the problem is that the base `/obj` is set to a certain plane. This does mean that sometimes, this VERY IMPORTANT marker is covered up by rocks or other objects. So, this seeks to get rid of that potential flaw, as well as do some other things.

Sometimes, when objects are rendered via GAGS or other code-means, they tend to have the same default Purple/White Sprite that any object without a valid icon_state has. This has caused me some confusion, so I have decided to create a new icon for conflict markers. This icon was designed to be as ugly as possible, while creating as much contrast as possible with the background by incorporating several colors into its design. I hope you find merge conflicts as unpleasant as I do.

I also updated mapmerge2 to have it so you can set the specified path of the object, as well as a small comment on the warnings if you do not heed it. I'm keeping the fact that mapmerge2 adds a name to the object just in case someone else really needs that. I also updated the linters to check for this path as well (even though the name and description should suffice for linting), and it should all be gravy from here.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Chiefly, it's fucking crap that the merge conflict markers aren't always visible, they can hide under rocks like this test case:

![image](https://user-images.githubusercontent.com/34697715/176027932-868ea019-0ab4-4a3d-9f3a-df8dd1dd8c20.png)

Which is terrible for mapping since you can't ever be certain where something is. You can use the search function in StrongDMM for the pre-fab ID of the merge conflicted item to eventually find the busted tile, but it's not like it's still obvious again.

I tested the greps and mapmerge, and everything appears to work plenty fine. Here's an example from a few test cases I ran:

![image](https://user-images.githubusercontent.com/34697715/176028133-f8f40302-2065-42ef-be35-5d575d265ee2.png)

![image](https://user-images.githubusercontent.com/34697715/176028550-52e8ff4e-2432-4fc2-8bf5-57983b8b9072.png)

![image](https://user-images.githubusercontent.com/34697715/176286150-ac67fb36-e13d-485a-a0ef-40f97a5a0329.png)

![image](https://user-images.githubusercontent.com/34697715/176286225-18305ba1-fbbc-42f4-9b06-f51a109ec38c.png)

Much better.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

Nothing that particularly concerns players.

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

san7890 committed:
Merge Conflict Markers - The Explicit Pathing

Hey there,

This PR corrects an issue I've been having with mapmerge2 these last few months. Basically, what it does is create a base `/obj` that is given the name `---Merge Conflict Marker---`. This is fine and all, but the problem is that the base `/obj` is set to a certain plane. This does mean that sometimes, this VERY IMPORTANT marker is covered up by rocks or other objects. So, this seeks to get rid of that potential flaw, as well as do some other things.

Sometimes, when objects are rendered via GAGS or other code-means, they tend to have the same default Purple/White Sprite that any object without a valid icon_state has. This has caused me some confusion, so I have decided to create a new icon for conflict markers. This icon was designed to be as ugly as possible, while creating as much contrast as possible with the background by incorporating several colors into its design. I hope you find merge conflicts as unpleasant as I do.

I also updated mapmerge2 to have it so you can set the specified path of the object, as well as a small comment on the warnings if you do not heed it. I'm keeping the fact that mapmerge2 adds a name to the object just in case someone else really needs that. I also updated the linters to check for this path as well (even though the name and description should suffice for linting), and it should all be gravy from here.

san7890 committed:
Adds further contrast to the DMI.

Fikou commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -732,3 +731,12 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+/obj/effect/mapping_helpers/merge_conflict_marker
+	name = "Merge Conflict Marker - DO NOT USE"
+	icon_state = "merge_conflict_marker"
+	desc = "If you are seeing this in-game: someone REALLY, REALLY, REALLY fucked up. They physically mapped in a fucking Merge Conflict Marker. What the shit."
```
why is this a mapping helper? someone will not be able to see this and report an error because mapping helpers immediately delete themselves

san7890 replied:
i might have an idea for this then

san7890 committed:
Splits the merge conflict marker into a generic /obj

I also added a thing where if it didn't get caught by linters and it showed up on Initalize, it would error to mapping logs, spit out an error in world, and do all sorts of stuff to remind you.

san7890 committed:
python new line

san7890 committed:
forgot to add a tab

timothymtorres commented:
This is great San.  The original conflict marker was easy to miss and this new one should be more eye catching.  +1

SuperNovaa41 reviewed (approved):
codes fine

Imaginos16 reviewed (approved):
yea

san7890 commented on tools/mapmerge2/merge_driver.py:
```diff
@@ -75,9 +75,13 @@ def three_way_merge(base, left, right):
+            obj_path = "/obj/merge_conflict_marker"
+            obj_name = "---Merge Conflict Marker---"
+            obj_desc = "A best-effort merge was performed. You must resolve this conflict yourself (manually) and remove this object once complete."
+            merged_movables = left_movables + [f'{obj_path}{{name = "{obj_name}",\n\tdesc = "{obj_desc}"}}'] + right_movables
```
this line was bugged, fixed in https://github.com/tgstation/tgstation/pull/68536
